### PR TITLE
feat: add gas optimize for CrossChain and TokenHub

### PR DESCRIPTION
**Pull request by cosinlink**
### Description

add gas optimize for CrossChain and TokenHub, replace rlp encode/decode with abi.encode/decode

### Rationale
for gas optimization

avg gas used of transferOut encoding : 
1. current:  15000 gas
2. new: 1435 gas

avg gas used of decodeTransferOutAckPackage:
1. current: 6289
2. new: 471

origin test:

```
# install dependency
# set `BSC_TESTNET_RPC=<your bsc testnet rpc>` to your .env
forge test --match-contract TokenHubTest  --gas-report -vvv

Running 6 tests for foundry-tests/TokenHub.t.sol:TokenHubTest
[PASS] test_rlp_fuzzy_test_case_1(uint256,address,address) (runs: 10000, μ: 15502, ~: 14725)
[PASS] test_rlp_fuzzy_test_case_1_abi_decode(uint256,address,address) (runs: 10000, μ: 1435, ~: 1435)
[PASS] test_rlp_fuzzy_test_case_2(uint256,address,uint32) (runs: 10000, μ: 16302, ~: 16243)
[PASS] test_rlp_fuzzy_test_case_2_abi_decode(uint256,address,uint32) (runs: 10000, μ: 1707, ~: 1707)
```

### Changes

Notable changes:
* add gas optimize for CrossChain and TokenHub, replace rlp encode/decode with abi.encode/decode
* ...

### CODE REVIEW

File: contracts/CrossChainV2.sol [hunk 1]
```diff
@@ -0,0 +1,587 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
```
Comment (guagualvcha) [inline contracts/CrossChainV2.sol:1]: Is this file necessary, as it is identical with the `CrossChain.sol`

File: contracts/middle-layer/TokenHubV2.sol [hunk 2]
```diff
@@ -0,0 +1,298 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
```
Comment (guagualvcha) [inline contracts/middle-layer/TokenHubV2.sol:2]: I expected more gas profile testing on the resource mirror, especially the grouphub.

The current implementation is not complete.

File: contracts/Config.sol [hunk 3]
```diff
@@ -14,15 +14,15 @@ abstract contract Config {
...
-    address public constant PROXY_ADMIN = address(0);
-    address public constant GOV_HUB = address(0);
-    address public constant CROSS_CHAIN = address(0);
-    address public constant TOKEN_HUB = address(0);
-    address public constant LIGHT_CLIENT = address(0);
-    address public constant RELAYER_HUB = address(0);
-    address public constant BUCKET_HUB = address(0);
-    address public constant OBJECT_HUB = address(0);
-    address public constant GROUP_HUB = address(0);
+    address public constant PROXY_ADMIN = 0x052008988d3b1023599aa61A852bf1c06b776159;
```
Comment (Pythonberg1997) [inline contracts/Config.sol:17]: Unnecessary changes

**Review by Pythonberg1997 — APPROVED** [review]

File: contracts/middle-layer/resource-mirror/encode/CmnEncode.sol [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    /*----------------- encode -----------------*/
+    function encodeCmnMirrorAckPackage(CmnMirrorAckPackage calldata mirrorAckPkg) external pure returns (bytes memory) {
+        return wrapEncode(TYPE_MIRROR, abi.encode(mirrorAckPkg));
+    }
+
+    function encodeCmnDeleteSynPackage(CmnDeleteSynPackage calldata synPkg) external pure returns (bytes memory) {
+        return wrapEncode(TYPE_DELETE, abi.encode(synPkg));
+    }
+
+    function encodeExtraData(ExtraData calldata _extraData) external pure returns (bytes memory) {
```
Comment (Pythonberg1997) [inline contracts/middle-layer/resource-mirror/encode/CmnEncode.sol:17]: This function seems useless anymore. abi.encode(_extraData) is enough.

File: contracts/middle-layer/resource-mirror/encode/CmnEncode.sol [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    function wrapEncode(uint8 opType, bytes memory msgBytes) public pure returns (bytes memory) {
+        return abi.encodePacked(opType, msgBytes);
+    }
+
+    /*----------------- decode -----------------*/
+    function decodeCmnMirrorSynPackage(
+        bytes calldata msgBytes
+    ) external pure returns (CmnMirrorSynPackage memory, bool success) {
```
Comment (Pythonberg1997) [inline contracts/middle-layer/resource-mirror/encode/CmnEncode.sol:28]: `success` in return value is useless

File: contracts/CrossChainV2.sol (hunk 1 above)
Comment (cosinlink) [reply contracts/CrossChainV2.sol]: removed

File: contracts/Config.sol (hunk 3 above)
Comment (cosinlink) [reply contracts/Config.sol]: fixed

**Comment by guagualvcha** [comment]
Please delete all the place than mention rlp

File: contracts/Deployer.sol [hunk 6]
```diff
@@ -113,9 +113,9 @@ contract Deployer {
         objectToken = addrs[12];
         groupToken = addrs[13];
         memberToken = addrs[14];
-        bucketRlp = addrs[15];
-        objectRlp = addrs[16];
-        groupRlp = addrs[17];
+        BucketEncode = addrs[15];
```
Comment (guagualvcha) [inline contracts/Deployer.sol:116]: Why is this in capital letters?

File: contracts/interface/IBucketEncode.sol [hunk 7]
```diff
@@ -4,7 +4,7 @@ pragma solidity ^0.8.0;
 
 import "../middle-layer/resource-mirror/storage/BucketStorage.sol";
 
-interface IBucketRlp {
+interface IBucketEncode {
     function decodeCmnCreateAckPackage(
```
Comment (guagualvcha) [inline contracts/interface/IBucketEncode.sol:8]: It make no sense that `decodeCmnxxx` is a function of `IBucketEncode`, I will suggest `IBucketCodec`

File: contracts/interface/IGroupEncode.sol [hunk 8]
```diff
@@ -4,7 +4,7 @@ pragma solidity ^0.8.0;
 
 import "../middle-layer/resource-mirror/storage/GroupStorage.sol";
 
-interface IGroupRlp {
+interface IGroupEncode {
```
Comment (guagualvcha) [inline contracts/interface/IGroupEncode.sol:7]: so is here

File: contracts/interface/IObjectEncode.sol [hunk 9]
```diff
@@ -4,7 +4,7 @@ pragma solidity ^0.8.0;
 
 import "../middle-layer/resource-mirror/storage/ObjectStorage.sol";
 
-interface IObjectRlp {
+interface IObjectEncode {
```
Comment (guagualvcha) [inline contracts/interface/IObjectEncode.sol:7]: so is here

File: contracts/lib/CmnPkg.sol [hunk 10]
```diff
@@ -2,38 +2,17 @@
...
-                success = true;
-            } else {
-                break;
-            }
-            idx++;
-        }
-        return (ackPkg, success);
+    function decodeCommonAckPackage(bytes memory msgBytes) internal pure returns (CommonAckPackage memory ackPkg, bool) {
+        ackPkg = abi.decode(msgBytes, (CommonAckPackage));
+        return (ackPkg, true);
```
Comment (guagualvcha) [inline contracts/lib/CmnPkg.sol:16]: what will happen if `msgBytes` can not decode to a  `CommonAckPackage`?

File: contracts/middle-layer/resource-mirror/BucketHub.sol [hunk 11]
```diff
@@ -7,18 +7,16 @@ import "@openzeppelin/contracts-upgradeable/utils/structs/DoubleEndedQueueUpgrad
...
+import "../../interface/IBucketEncode.sol";
 import "../../interface/IERC721NonTransferable.sol";
 
 contract BucketHub is BucketStorage, AccessControl, CmnHub, IBucketHub {
     using DoubleEndedQueueUpgradeable for DoubleEndedQueueUpgradeable.Bytes32Deque;
-    using RLPEncode for *;
-    using RLPDecode for *;
 
-    function initialize(address _ERC721_token, address _additional, address _bucketRlp) public initializer {
+    function initialize(address _ERC721_token, address _additional, address _bucketEncode) public initializer {
```
Comment (guagualvcha) [inline contracts/middle-layer/resource-mirror/BucketHub.sol:16]: Let us not user `codec` as an address.

File: contracts/middle-layer/resource-mirror/encode/CmnEncode.sol (hunk 5 above)
Comment (cosinlink) [reply contracts/middle-layer/resource-mirror/encode/CmnEncode.sol]: fixed

File: contracts/middle-layer/resource-mirror/encode/CmnEncode.sol (hunk 4 above)
Comment (cosinlink) [reply contracts/middle-layer/resource-mirror/encode/CmnEncode.sol]: fixed

File: contracts/middle-layer/resource-mirror/BucketHub.sol (hunk 11 above)
Comment (cosinlink) [reply contracts/middle-layer/resource-mirror/BucketHub.sol]: fixed

File: contracts/lib/CmnPkg.sol (hunk 10 above)
Comment (cosinlink) [reply contracts/lib/CmnPkg.sol:16]: reverted

File: contracts/interface/IGroupEncode.sol (hunk 8 above)
Comment (cosinlink) [reply contracts/interface/IGroupEncode.sol]: fixed

File: contracts/interface/IBucketEncode.sol (hunk 7 above)
Comment (cosinlink) [reply contracts/interface/IBucketEncode.sol]: fixed

File: contracts/Deployer.sol (hunk 6 above)
Comment (cosinlink) [reply contracts/Deployer.sol]: removed

**Review by Pythonberg1997 — APPROVED** [review]

**Review by yutianwu — APPROVED** [review]

**Review by guagualvcha — APPROVED** [review]
